File: Web/src/domain/instrument/breath.ts
import type { PositionOnHarmonica } from '../playing/positionOnHarmonica.js'
import { topmostOf } from '../playing/positionOnHarmonica.js'

export type Breath = 'blow' | 'draw'

const slackAcrossTheBoundary = 0.06

export function reversed(breath: Breath): Breath {
    return breath === 'blow' ? 'draw' : 'blow'
}

export function isCrossingTheBreathBoundary(position: PositionOnHarmonica): boolean {
    return Math.abs(position.fractionAboveCentreLine) < slackAcrossTheBoundary
}

export function breathOfTheTopmost(positions: readonly PositionOnHarmonica[]): Breath | null {
    const topmost = topmostOf(positions)
    if (topmost === null) return null

    return topmost.fractionAboveCentreLine >= 0 ? 'blow' : 'draw'
}

File: Apps/iOS/Domain/Entities/Instrument/Breath.swift
enum Breath {
    case blow
    case draw

    var reversed: Breath {
        switch self {
        case .blow: .draw
        case .draw: .blow
        }
    }
}

// MARK: - PositionOnHarmonica + Breath

extension PositionOnHarmonica {
    private static let slackAcrossTheBoundary = 0.06

    var isCrossingTheBreathBoundary: Bool {
        abs(fractionAboveCentreLine) < Self.slackAcrossTheBoundary
    }
}

// MARK: - Breath + PositionOnHarmonica

extension Breath {
    init?(topmostOf positions: [PositionOnHarmonica]) {
        guard let topmost = PositionOnHarmonica.topmost(of: positions) else { return nil }

        self.init(at: topmost)
    }

    private init(at position: PositionOnHarmonica) {
        self = position.fractionAboveCentreLine >= 0 ? .blow : .draw
    }
}
